### ISSUE TITLE
aws_iam_role data source doesn't support filtering by tags

### ISSUE THREAD

**Issue by vir-dis** [terraform-providers/terraform-provider-aws#7419]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The main data source [documentation](https://www.terraform.io/docs/configuration/data-sources.html) says data sources can be filtered by attributes or tags, but the [aws_iam_role](https://www.terraform.io/docs/providers/aws/d/iam_role.html) data source doesn't support any filtering.

We would like to be able to retrieve a list of iam_roles based on tags.

### New or Affected Resource(s)

* aws_iam_role

### Potential Terraform Configuration

```hcl
data "aws_iam_role" "web_roles" {
  filter {
    name   = "tag:Component"
    values = ["web"]
  }
}

```

### References

* #6596

**Comment by Bharathkumarraju** [comment]
is this implemented already?

**Comment by marzigolbaz** [comment]
This does not seem to be implemented. Getting error:
```
Blocks of type "filter" are not expected here.
```
When this feature is going to get added?

**Comment by nitrocode** [comment]
Please correct me if I'm wrong but I don't believe the AWS API has a way to filter roles or users by tags and so this ticket is blocked unless they listed all the iam roles, grabbed each role's tags, then filtered all of those which would be A LOT of API calls.

If you're interested in this, you could write up a support request with AWS asking to be able to programatically get iam roles and users using tags.

**Comment by cnelissen** [comment]
There is an aws cli command that can grab all roles, including tags, so the API has to support this type of query.

Example output:

```
Macbook % aws iam list-roles

{
    "Roles": [
        {
            "Path": "/aws-reserved/sso.amazonaws.com/",
            "RoleName": "AWSReservedSSO_AdministratorAccess_ef8f767c88a9",
            "RoleId": "AROATWYDUPRSZOMC",
            "Arn": "arn:aws:iam::######:role/aws-reserved/sso.amazonaws.com/AWSReservedSSO_AdministratorAccess_ef8f767c88a9",
            "CreateDate": "2021-03-01T15:01:50+00:00",
            "AssumeRolePolicyDocument": {
                "Version": "2012-10-17",
                "Statement": [
                    {
                        "Effect": "Allow",
                        "Principal": {
                            "Federated": "arn:aws:iam::######:saml-provider/AWSSSO_43578fc0f4b5_DO_NOT_DELETE"
                        },
                        "Action": [
                            "sts:AssumeRoleWithSAML",
                            "sts:TagSession"
                        ],
                        "Condition": {
                            "StringEquals": {
                                "SAML:aud": "https://signin.aws.amazon.com/saml"
                            }
                        }
                    }
                ]
            },
            "MaxSessionDuration": 43200,
            "Tags": []
        }
...
 ```

**Comment by joanayma** [comment]
Will this ever be considered?

**Comment by marcdavoli** [comment]
You should be able to accomplish this using the `aws_iam_roles` Data Source, more info here : 
https://github.com/hashicorp/terraform-provider-aws/issues/14470

**Comment by max-rocket-internet** [comment]
> You should be able to accomplish this using the aws_iam_roles Data Source, more info here :
> https://github.com/hashicorp/terraform-provider-aws/issues/14470

This data resource does not support filtering by tags.

**Comment by nitrocode** [comment]
Yes exactly, it only supports a `name_regex` or a `path_prefix` and both names and prefixes are immutable.

**Comment by terenho** [comment]
The AWS SDK for [ListRoles](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/iam#Client.ListRoles) does [not support tag input](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/iam#ListRolesInput). It only takes `PathPrefix *string`

**Comment by terenho** [comment]
@cnelissen Even though `aws iam list-roles` returns a `Tags` output, it is always empty. In https://awscli.amazonaws.com/v2/documentation/api/2.18.18/reference/iam/list-roles.html, it says

> **Note**:
>
> IAM resource-listing operations return a subset of the available attributes for the resource. This operation does not return the following attributes, even though they are an attribute of the returned object:
> 
>  - PermissionsBoundary
>  - RoleLastUsed
>  - Tags
>
> To view all of the information for a role, see GetRole .

**Comment by nitrocode** [comment]
What about the GetResources API? Could that be used here?

```bash
aws resourcegroupstaggingapi get-resources \
    --resource-type-filters role \
    --tag-filters Key=Environment,Values=Production \
    --tags-per-page 100
```

`role` is supported

- https://docs.aws.amazon.com/cli/latest/reference/resourcegroupstaggingapi/get-resources.html
- https://docs.aws.amazon.com/resourcegroupstagging/latest/APIReference/supported-services.html

Supporting a generic data source for retrieving tagged resources would be even more awesome as it would be a good escape hatch for existing data sources that cannot support it.

**Comment by terenho** [comment]
> ```shell
> --tag-filters Key=
> ```

That's a no. I tried
```
aws resourcegroupstaggingapi get-resources \
    --resource-type-filters iam:role \
    --tag-filters Key=Environment,Values=Production \
    --tags-per-page 100
```
and it always returns an empty list.

Reading https://docs.aws.amazon.com/resourcegroupstagging/latest/APIReference/supported-services.html more carefully. The `GetResources` does not support `iam:role`
![image](https://github.com/user-attachments/assets/92ae809e-63b2-4920-b697-24879f579084)
